prow/plugins/trigger/generic-comment.go

```diff
@@ -141,36 +141,44 @@ type GitHubClient interface {
 // If a comment that we get matches more than one of the above patterns, we
 // consider the set of matching presubmits the union of the results from the
 // matching cases.
-func FilterPresubmits(honorOkToTest bool, gitHubClient GitHubClient, body string, pr *github.PullRequest, presubmits []config.Presubmit) ([]config.Presubmit, error) {
+func FilterPresubmits(honorOkToTest bool, gitHubClient GitHubClient, body string, pr *github.PullRequest, presubmits []config.Presubmit) ([]config.Presubmit, []config.Presubmit, error) {
 	org, repo, sha := pr.Base.Repo.Owner.Login, pr.Base.Repo.Name, pr.Head.SHA
 	filter, err := presubmitFilter(honorOkToTest, gitHubClient, body, org, repo, sha)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	return filterPresubmits(filter, gitHubClient, pr, presubmits)
 }
 
-// filterPresubmits determines which presubmits should run by evaluating the
-// user-provided filter.
-func filterPresubmits(filter filter, gitHubClient GitHubClient, pr *github.PullRequest, presubmits []config.Presubmit) ([]config.Presubmit, error) {
+// filterPresubmits determines which presubmits should run and which should be skipped
+// by evaluating the user-provided filter.
+func filterPresubmits(filter filter, gitHubClient GitHubClient, pr *github.PullRequest, presubmits []config.Presubmit) ([]config.Presubmit, []config.Presubmit, error) {
 	org, repo, number, branch := pr.Base.Repo.Owner.Login, pr.Base.Repo.Name, pr.Number, pr.Base.Ref
 	changes := config.NewGitHubDeferredChangedFilesProvider(gitHubClient, org, repo, number)
 	var filteredPresubmits []config.Presubmit
+	filteredContexts := sets.NewString()
+	var skippedPresubmits []config.Presubmit
 	for _, presubmit := range presubmits {
 		matches, forced, defaults := filter(presubmit)
 		if !matches {
 			continue
 		}
 		shouldRun, err := presubmit.ShouldRun(branch, changes, forced, defaults)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 		if shouldRun {
 			filteredPresubmits = append(filteredPresubmits, presubmit)
+			filteredContexts.Insert(presubmit.Context)
+		} else if !filteredContexts.Has(presubmit.Context) {
```

Comment:
This is incorrectly dependent on the order that presubmits are defined in `presubmits`.
Consider the case in which we are responding to an event for a PR on branch A and we have a branch sharded job with versions for branch B and branch A (in that order). This will incorrectly post a skipped status for branch B before triggering the branch A job shard. Please add a test case for this.

I think it would be cleanest to build a slice of jobs that might need to be skipped and then remove any jobs with a context that appears in `filteredContexts` after *all* the filtered contexts have been identified.